test/src/test/java/io/jenkins/jenkinsfile/runner/JenkinsTestEmbedder.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2004-2018, Sun Microsystems, Inc., Kohsuke Kawaguchi, Erik Ramfelt,
+ * Yahoo! Inc., Tom Huybrechts, Olivier Lamy, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package io.jenkins.jenkinsfile.runner;
+
+import io.jenkins.jenkinsfile.runner.util.JenkinsRecipe;
+import io.jenkins.jenkinsfile.runner.util.LenientRunnable;
+import io.jenkins.jenkinsfile.runner.util.TestEnvironment;
+import org.junit.runner.Description;
+
+import java.lang.annotation.Annotation;
+
+/**
+ * JenkinsEmbedder used in TestEnvironment
+ */
+public abstract class JenkinsTestEmbedder extends JenkinsEmbedder {
+
+    protected Description testDescription;
+
+    /**
+     * Number of seconds until the test times out.
+     */
+    public int timeout = Integer.getInteger("jenkins.test.timeout", System.getProperty("maven.surefire.debug") == null ? 180 : 0);
+
+    public JenkinsEmbedder with(Description testDescription) {
+        this.testDescription = testDescription;
+        env = new TestEnvironment(testDescription);
+        return this;
+    }
+
+    /**
+     * Called during the {@link #before()} to give a test case an opportunity to
+     * control the test environment in which Hudson is run.
+     */
+    @Override
+    public void recipe() throws Exception {
```

Comment:
Weren't recipes used previously to setup the Jenkins environment where to run the Pipeline?

Reply:
No class had the annotation `@JenkinsRecipe`, so they were not previously used at all. They can be used that way, but it's not actually used.